add conditional statements to Link()

mtremmel opened the pull request:
Make it so the user can put their own constraints on Link in addition to choosing the max/min of a given property. Also, now the property used for choosing the correct link can be a live calculated property rather than just stored properties. Can take any number of additional constraints, though still requires a determiner property and max/min.

apontzen commented on tangos/live_calculation/builtin_functions/link.py:
```diff
@@ -20,14 +20,18 @@ def link(*args):
 
 link.set_input_options(0, provide_proxy=True, assert_class=StoredProperty)
-link.set_input_options(1, provide_proxy=True, assert_class=StoredProperty)
+link.set_input_options(1, provide_proxy=True)
```
I can't remember exactly why the class of 1 had to be `StoredProperty` but equally is there a clear reason for removing that requirement?

Also I am concerned that since the other N arguments don't have `provide_proxy=True` the code may end up trying to fetch the values of say `BH_mass` before actually calling the `link` function. Probably won't break anything but could be wasteful.

In other words we might want something like
```python
link.set_input_options(2, assert_class=FixedInput)
for i in range(20): # TODO: implement a way to set provide_proxy on all possible arguments, to remove arbitrary upper limit
    link.set_input_options(i, provide_proxy=True)
```

It's ugly but should make things more efficient

mtremmel replied:
Ok. I agree... a bit ugly, but I see your point

apontzen replied:
Third input is the string 'max' or 'min' no?

Yes I agree about the first input and also it should presumably be `range(3,20)` rather than `range(20)` which would overwrite the options for 0,1,2

apontzen commented on tangos/live_calculation/builtin_functions/link.py:
```diff
@@ -20,14 +20,18 @@ def link(*args):
-        internal_getter = Link(link_getter, MultiCalculation(ReturnInputHalos(), property_getter))
+        if constraints:
+            internal_getter = Link(link_getter, MultiCalculation(ReturnInputHalos(), property_getter,*constraints))
+            internal_getter.set_constraints_columns(list(np.arange(len(constraints))+2))
```
is there a reason this has to be a list? Could just keep the numpy array?

mtremmel replied:
Yeah I suppose so

apontzen commented:
Two more thoughts

* Oops there is also a problem with the existing tests, looks like it is a numpy version specific thing:
https://travis-ci.org/pynbody/tangos/jobs/364237089

* Could you update `docs/live_calculation.md`?

mtremmel commented:
As I am writing new checks I realized that there is an issue... the constraint fails when the calculation has returned None (it comes up as True when, because the object doesn't have one or more of the properties used in the constraint, it should be False). Trying to get around this.... strangely not as obvious as I would have thought.

apontzen commented:
It might make sense to fold this into the `QueryMask` that `QueryMultivalueFolding` creates... just a thought. As well as solving your problem, it could also get rid of the ugly expressions like `np.where(truth_slice)[0][determiner_slice[truth_slice].argmax()]`...

mtremmel commented:
Ok so I've fixed the issue I was having before with linked objects not having data for a user defined constraint. I also now use masks... I'm unsure if I do it in the best way? I create a new mask object that uses a new function to mask on False rather than None values. I've also added some documentation in `live_calculation.md` for this functionality and `link()` in general.

apontzen commented on tangos/live_calculation/query_masking.py:
```diff
@@ -26,6 +26,24 @@ def mark_nones_as_masked(self, input):
+            raise ValueError("Not able to use an input of this shape to determine a query mask")
+
+        assert mask.shape==(self.N,)
+        self.results_target = np.where(mask)
```
This seems to be duplicating the exact same code from above. Maybe make a generic method like `_retain_row_if_input_satisfies(self, input, condition)` then `mark_false_as_masked` could call `_retain_row_if_input_satisfies(self, input, is_not_false)` and `mark_none_as_masked` could call `_retain_row_if_input_satisfies(self, input, is_not_none)`

mtremmel replied:
good idea I can do that!

apontzen commented on tangos/live_calculation/query_multivalue_folding.py:
```diff
@@ -34,11 +35,25 @@ def refold(self, results_to_refold):
+
+            if len(determiner_slice)!=0 and True in truth_slice:
+                mask2 = QueryMask()
+                mask2.mark_false_as_masked(truth_slice)
```
This all seems quite complicated - could one actually just use the same mask object twice, i.e. allow the mask to be 'updated'?

mtremmel replied:
so really the only difference is no reinitialization of a new mask object and just do `mask.mark_fasle_as_masked(truth_slice)`, which would reset and create essentially a new mask because truth_slice is not the same length as the original masked array. Do I have that right?

apontzen replied:
Yes, that's right... as currently implemented marking the mask loses the previous mask but that could be fixed so that they get &ed together

mtremmel replied:
hmmm yeah. I suppose that I would also have to make sure that the original `truth_slice` is also the same size as originally masked array. As it is now, truth_slice is initialized based on the first mask. But I can see how this is a cleaner way of doing it. I'll see if I can change the mask to be more versatile

mtremmel commented:
Changes made to the QueryMask class to make it simpler and also cumulative so that only a single mask object needs to be used when refolding based on multiple conditions

apontzen commented on tangos/live_calculation/query_masking.py:
```diff
@@ -1,31 +1,43 @@
+
+    def mark_nones_as_masked(self, input):
+        """Mark any rows in the input that are None as masked, excluding them from future queries."""
+        self._retain_row_if_true(input, is_not_none)
```
Can I make an annoying suggestion that you rename `_retain_row_if_true` to `_retain_row_if`? I think this would be clearer (otherwise the bit below about `is_not_false` becomes v confusing!)
